### PR TITLE
[5.1] Improvements to Redis cache tagging

### PR DESCRIPTION
**Pull request by mstephens**
Use SETs rather than LISTs for storing cache key references.

### CODE REVIEW

File: src/Illuminate/Cache/RedisTaggedCache.php [hunk 1]
```diff
@@ -143,8 +143,9 @@ protected function deleteKeysByReference($reference)
      */
     protected function deleteValues($referenceKey)
     {
-        $values = array_unique($this->store->connection()->lrange($referenceKey, 0, -1));
+        $values = array_unique($this->store->connection()->smembers($referenceKey));
 
+        // Remove each tagged cache item
```
Comment (GrahamCampbell) [inline src/Illuminate/Cache/RedisTaggedCache.php]: this comment is unnecessary

Comment (mstephens) [inline src/Illuminate/Cache/RedisTaggedCache.php]: @GrahamCampbell Yes I have done, perhaps a bit more friendly with your messages. I am only helping to contribute!

**Comment by taylorotwell** [comment]
Needs actual description of how this is an improvement, what bugs it fixes, etc.

**Comment by mstephens** [comment]
@taylorotwell it prevents multiple child tagged cache key references being stored against the tag key type (standard or forever). When a cache tag is flushed this is how the child keys are referenced and we know which ones to flush.

Currently it can store multiple, and although it works fine with duplicates, is inefficient/extra overhead.

There is no functional change to the end user, just a cache flush required to change from using SET storage to LIST.

**Comment by taylorotwell** [comment]
Thanks

**Comment by crynobone** [comment]
This PR actually broke https://github.com/tymondesigns/jwt-auth refresh token code. 

screen shot 2016-03-28 at 10 51 58 am

> There is no functional change to the end user, just a cache flush required to change from using SET storage to LIST.

The code was run on a vagrant instance with no existing cache, and it still getting the same error code. Furthermore, saying that it would work if you flush the cache on a patch release is still a breaking change!

**Comment by taylorotwell** [comment]
Sigh. I'll revert it in the morning. I've gotta learn to stop merging things that try to fix something that isn't broken in the first place.

**Comment by mstephens** [comment]
Hi short for time at the moment but... the old method was broken. Leaving stray and duplicate keys in the tag reference store. When you're storing lots of tags it becomes more problematic. Also before 5.2.14 tagging on standard keys didn't work at all...

The reason for these errors are people are not flushing the cache after a composer update (something you should do in my opinion).

I see no way to make these changes like this without doing so. Should an artisan cache clear not run as part of the composer deployment ? It works fine as soon as this is followed and all tests run as they should. 

I don't see how else these changes could be handled. Without try/catches throughout the tagging code...

:(

**Comment by crynobone** [comment]
Clearing the cache used for JWT authentication is not wise in my opinion, this would mean that on every deployment the cache would be cleared and users accessing the mobile app would need to be re-authenticated (login in by key-in the username and password). 

BUT MOST IMPORTANTLY, as I said before this also happen on a newly up Vagrant instance with NO EXISTING cache.

**Comment by mstephens** [comment]
Hi short for time at the moment (vacation) but... the old method (there from 5.1) was broken. Leaving stray and duplicate keys in the tag reference store... Not detrimental but not how it should have been. When you're storing lots of tags it becomes more problematic. In terms of overall, before my previous fix tagging on standard keys didn't work at all...

The reason for these errors are people are not flushing the cache after a composer update (something you should do in my opinion).

I see no way to make these changes like this without doing so. Should an artisan cache clear not run as part of the composer deployment ? It works fine as soon as this is followed and all tests run as they should. 

I don't see how else these changes could be handled. Without try/catches throughout the tagging code...

:(

**Comment by mstephens** [comment]
@crynobone which method are you receiving errors on? Tried on a vm and all ok. Using a shared cache store?

**Comment by crynobone** [comment]
Did you even read/look at the error screenshot that I posted?

**Comment by GrahamCampbell** [comment]
This should be reverted, and moved to 5.3 imo.

**Comment by mstephens** [comment]
Yeah I saw the screenshot. I asked which php method... A flush or a push? 

As for 4. I am not saying it was, I mentioned what needed to happen in the commit message. I applied the patch to the correct branch in 5.1 and up to the repo owner what happens. 

I am at a loss as to how a change like this would be introduced in a different way... 

Caching was never tested fully in 5.2,  if it had been cache tags would have flushed from launch, forever or standard ones (only forever used to). No need for the tone... just trying to help and fix a long running bug, alongside a full time job :)

**Comment by taylorotwell** [comment]
Should be moved to 5.3.

**Comment by crynobone** [comment]
> Also it would be a fair assumption to flush cache after a framework upgrade.

The assumption that everyone would flush cache on every single patch releases is just like saying "I guess this code would work fine, it works for my use case".

I agree that It would be fair, for a major release but not on patch release (as you might be deploying code on a daily basis). But again, it up to developer to choose how they want to proceed during the upgrade as noted on the upgrade note for 5.3. Adding the changes in the middle of patch release without any upgrade note (there shouldn't be any) is bad. I had to revert our planned deployment to production last monday because it suddenly throwing errors where it shouldn't happen, and had to look through commit history to find the cause of this error not only on `tymon/jwt-auth` repository but also `laravel/framework`. So yes I'm deeply frustrated with the numbers of PRs that took breaking changes very lightly.

**Comment by boynet** [comment]
just a question related to this.
is there a way to make tests capture this kind of bugs between patch updates?

**Comment by mstephens** [comment]
@crynobone I see what you're saying, but in principal (not saying in just Laravel's case) cache should never be treated as portable between versions for a framework. If you updated your system from 5.2.23 to 5.2.24 you would always check after the update and run any tests/cache clears, these are then run the same on live too. 

Many apps (such as Facebook) change the way in which cache is stored between revisions, it's just these are run as part of the upgrade. The problem with the composer deployment of Laravel is these are bypassed. I did mention in the pull request what needed to be done, but I'm not in control on the release notes. Perhaps it could have been highlighted then by someone reviewing.  

Yes, this should probably have been merged into 5.3, I followed the contribution guidelines and plugged it from where the issue became present.

@boynet I looked at this when making changes now (and previously), currently the cache tag tests are very minimal and use mockery to test the method calls receive the correct data. I guess perhaps some kind of actual simulation of data would help for third party services, although how you'd check between revisions in a test would be interesting. To be honest I think the CachedTag methods could do with a little re-factoring. I also see no reason why tagging could not be present in file based and database back-ends too... I'll try to look should I get time.

**Comment by crynobone** [comment]
> If you updated your system from 5.2.23 to 5.2.24 you would always check after the update and run any tests/cache clears

We have tests on our application that test our code, and we trust Laravel and all the others packages (that we carefully pick) to respect the non-breaking changes. Now, do you know exactly what has changed between v5.2.23 to v5.2.24 without having to manually read https://github.com/laravel/framework/compare/v5.2.23...v5.2.24 or https://laraver.xyz/versions/v5.2.24

Laravel doesn't provide any CHANGELOG between patch release. (I don't mind this as Laravel promise no breaking changes on patch release).

>  I did mention in the pull request what needed to be done, but I'm not in control on the release notes. Perhaps it could have been highlighted then by someone reviewing.

It were mention only once asked by @taylorotwell, don't you think it should be mention on the first description? wouldn't that means anyone who come across the PR would notice it more? And furthermore what's with the PR title **Improvements to Redis cache tagging**, that doesn't even sound like fixing a bug at all.

**Pull request merged**